hub/util/index.py

```diff
@@ -0,0 +1,89 @@
+from typing import Union
+
+
+def merge_slices(existing_slice: slice, new_slice: slice) -> slice:
+    """Compose two slice objects
+
+    Given an iterable x, the following should be equivalent:
+    x[existing_slice][new_slice]
+    x[merge_slices(existing_slice, new_slice)]
+
+    Args:
+        existing_slice (slice): The existing slice to be restricted.
+        new_slice (slice): The new slice to be applied to the existing slice.
+
+    Returns:
+        slice: the composition of the given slices
+    """
+
+    # Combine the steps
+    step1 = existing_slice.step if existing_slice.step is not None else 1
+    step2 = new_slice.step if new_slice.step is not None else 1
+    step = step1 * step2
+
+    # Combine the start points
+    start1 = existing_slice.start if existing_slice.start is not None else 0
+    start2 = new_slice.start if new_slice.start is not None else 0
+    start = start1 + start2 * step1
+
+    # Combine the end points
+    stop1 = existing_slice.stop
+    stop2 = new_slice.stop
+
+    if stop2 is None:
+        stop = stop1
+    else:
+        stop = start + (stop2 - start2) * step1
+        if stop1 is not None:
+            stop = min(stop, stop1)
+
+    return slice(start, stop, step)
+
+
+class Index:
+    def __init__(self, item: Union[int, slice, "Index"] = None):
+        """Create a new Index from an int, slice, or another Index."""
+        if item is None:
+            item = slice(None)
+
+        if isinstance(item, Index):
```

Comment:
Can we not use "item" everywhere? Passing item as argument and accessing attributes like item.item seems really unusual to me.
Maybe we can rename the class Index to HubIndex and use "index" instead of "item" everywhere?

Reply:
Yeah, I don't really like it either. It's going to change when I update for sequences, so I'll leave this in for now.

Reply:
Please add a TODO for this.